File: app/src/main/java/com/chichiangho/base/http/ProgressRequestBody.java
package com.chichiangho.base.http;

/**
 * Created by chichiangho on 2017/5/9.
 */

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.Buffer;
import okio.BufferedSink;
import okio.ForwardingSink;
import okio.Okio;
import okio.Sink;

public class ProgressRequestBody extends RequestBody {
    private final RequestBody requestBody;
    private final OkHttpClient.ProgressListener progressListener;
    private BufferedSink bufferedSink;

    public ProgressRequestBody(RequestBody requestBody, OkHttpClient.ProgressListener progressListener) {
        this.requestBody = requestBody;
        this.progressListener = progressListener;
    }

    @Override
    public MediaType contentType() {
        return requestBody.contentType();
    }

    @Override
    public long contentLength() throws IOException {
        return requestBody.contentLength();
    }

    @Override
    public void writeTo(BufferedSink sink) throws IOException {
        if (bufferedSink == null) {
            bufferedSink = Okio.buffer(sink(sink));
        }
        requestBody.writeTo(bufferedSink);
        bufferedSink.flush();

    }

    private Sink sink(Sink sink) {
        return new ForwardingSink(sink) {
            long bytesWritten = 0L;
            long contentLength = 0L;

            @Override
            public void write(Buffer source, long byteCount) throws IOException {
                super.write(source, byteCount);
                if (contentLength == 0) {
                    contentLength = contentLength();
                }
                bytesWritten += byteCount;
                if (progressListener != null && bytesWritten != contentLength) {//不触发100%，交由callBack处理done和error，防止Rxjava的onComplete和onError同时触发导致crash
                    progressListener.onProgress((int) (bytesWritten * 100 / contentLength), null);
                }
            }
        };
    }
}

File: library_rx_okhttp/src/main/java/com/chichaingho/library_rx_okhttp/ProgressRequestBody.kt
package com.chichaingho.library_rx_okhttp

/**
 * Created by chichiangho on 2017/5/9.
 */

import okhttp3.MediaType
import okhttp3.RequestBody
import okio.*
import java.io.IOException

class ProgressRequestBody(private val requestBody: RequestBody, private val progressListener: OkHttpClient.ProgressListener?) : RequestBody() {
    private var bufferedSink: BufferedSink? = null

    override fun contentType(): MediaType? = requestBody.contentType()

    @Throws(IOException::class)
    override fun contentLength(): Long = requestBody.contentLength()

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        if (bufferedSink == null) {
            bufferedSink = Okio.buffer(sink(sink))
        }
        requestBody.writeTo(bufferedSink!!)
        bufferedSink?.flush()

    }

    private fun sink(sink: Sink): Sink {
        return object : ForwardingSink(sink) {
            internal var bytesWritten = 0L
            internal var contentLength = 0L

            @Throws(IOException::class)
            override fun write(source: Buffer, byteCount: Long) {
                super.write(source, byteCount)
                if (contentLength == 0L) {
                    contentLength = contentLength()
                }
                bytesWritten += byteCount
                if (progressListener != null && bytesWritten != contentLength) {//不触发100%，交由callBack处理done和error，防止Rxjava的onComplete和onError同时触发导致crash
                    progressListener.onProgress((bytesWritten * 100 / contentLength).toInt(), null)
                }
            }
        }
    }
}
